### ISSUE TITLE
core plugins decoupling: allow separate dependencies

### ISSUE THREAD

**Issue by gabor** [grafana/grafana#77998]
there is work ongoing on decoupling of the core datasources, and it would be great if they could have their own dependencies (javascript and go).

an example of why this is important:
- the postgres datasource plugin is relying on the https://github.com/lib/pq go module to handle postgres database connections.
- but, it's also used by grafana when it is configured to use postgres as it's internal database.
- so, if any change has to happen to this module (upgrade/downgrade/replace_with_other_module), it will affect two separate use-cases: datasource-plugin, grafana-internal-database. this means such changes have to be synchronized between different squads.

(this is not just a theoretical situation, the need to improve the https://github.com/lib/pq situation is real, considering it is in maintenance mode)

how decoupled dependencies would help here:
- faster changes, the datasource-squad can decide and execute the change by themselves, no need to synchronize with other squads
- more stable grafana, because just replacing the datasource-scenario's go module is a smaller change compared to replacing it for both scenarios (of course, the postgres-as-grafana-internal-database scenario's squad may decide to upgrade too, but that upgrade can be done separately, so one big change becomes two smaller changes)

**Comment by jackw** [comment]
> it would be great if they could have their own dependencies (javascript and go).

I cannot speak for the backend but a decoupled plugin should be a yarn workspace where frontend is concerned. As such it should already be able to define its own NPM javascript dependencies in the package.json. Maybe I'm missing something though. 🤔

**Comment by andresmgot** [comment]
Thanks for writing down the issue @gabor :+1: 

The tradeoff here (or downside) is that it requires more maintenance. There would be N new go.mod files that will require extra eyes to be kept up to date. For the frontend dependencies, this is mostly automated with Renovate + Lerna but I don't think there is tooling mature enough to do the same for Go packages (I may be wrong, it would require some investigation). The maintenance of the root go.mod is already a big burden so I'd be wary of adding more to that.

**Comment by gabor** [comment]
@jackw wrote:
> I cannot speak for the backend but where frontend is concerned a decoupled plugin should be a yarn workspace. As such it should already be able to define its own NPM javascript dependencies in the package.json.

oh, right, sorry, i was mostly focused on the `go` part and did not check the `javascript` part. you're right, there's for example https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/grafana-testdata-datasource/package.json

**Comment by gabor** [comment]
@andresmgot wrote:

> The tradeoff here (or downside) is that it requires more maintenance. There would be N new go.mod files that will require extra eyes to be kept up to date. For the frontend dependencies, this is mostly automated with Renovate + Lerna but I don't think there is tooling mature enough to do the same for Go packages (I may be wrong, it would require some investigation). The maintenance of the root go.mod is already a big burden so I'd be wary of adding more to that.

i see your point, but the way i'm looking at this, the maintenance burden would be split between core-grafana-maintainers and the plugin's maintainers.. i mean:
- the `go.mod` for the datasource-plugin would be the responsibility of the squad that owns the datasource, so there wouldn't be an increased burden on the maintainers of the "core" `go.mod` file
- also, this could make the "core" `go.mod` smaller, because the dependencies only used by the datasource-plugins, for example `golang.org/x/net` would be removed from the "core" `go.mod` file

WDYT?

**Comment by andresmgot** [comment]
yes, I am afraid this is a mono-repo / non-mono-repo situation. 

Right now other plugins get updates "for free" so it's difficult to reach a conclusion that fits 100% of the cases. We could also implement some kind of opt-in mechanism so people can decide to do it or not but that also complicates things implementation-wise.

**Comment by tolzhabayev** [comment]
We are going for externalization so this can be closed
